lib/controllers/flow/payment_flow_controller.dart

```diff
@@ -0,0 +1,166 @@
+      // is the signature.
+      _transactionRepository.setData(
+        transaction.id,
+        <String, dynamic>{
+          'authentication': {
+            'value': signature,
+          },
+          'responseType': ResponseType.authorized.toJsonString(),
+        },
+        merge: true,
+      );
+    } else {
+      // TODO(kkzeng): What happens when the authentication fails?
+    }
+  }
+
+  void _startListening(String id) {
+    _unsubscriber = _transactionRepository.listen(id, onValue: _onValue);
+  }
+
+  void _stopListening() {
+    _unsubscriber();
+  }
+
+  void _setAwaitingUpdate() {
+    // Update the payment status to be waiting for further update.
+    // This is intended to inform the UI that the user is not expected to
+    // make any action and just need to wait. For example, a circular progress
+    // indicator could be displayed.
+    isAwaitingUpdate = true;
+    update();
+  }
+
+  void _onValue(Transaction transaction) {
+    // Put the document id in the model object
+    transaction.id = documentId;
+
+    final oldValue = this.transaction;
+    this.transaction = transaction;
+
+    switch (transaction.status) {
+      case TransactionStatus.pendingPartyLookup:
+        // Do nothing here as we just need to wait for the server
+        // to request a party lookup in Mojaloop.
+        break;
+
+      case TransactionStatus.pendingPayeeConfirmation:
+        // The server has finished performing a party lookup and now it is
+        // the time for us to ask the user whether the party information
+        // (e.g., name) is correct. If yes, the user needs to key in some
+        // other values (e.g, transaction amount) to continue the process.
+        if (oldValue.status == TransactionStatus.pendingPartyLookup) {
+          // The transaction data has been updated
+          isAwaitingUpdate = false;
+          update();
+
+          Get.to<dynamic>(PaymentConfirmation(this));
+        }
+        break;
+
+      case TransactionStatus.authorizationRequired:
+        // The user must provide its authorization to continue the transaction.
+        // Here, the app has received the quote of the transaction from the server.
+        // The user will be able to get information about the fee and commission
+        // that are imposed by the FSPs and also more details about the transfer amount.
+        if (oldValue.status == TransactionStatus.pendingPayeeConfirmation) {
+          // The transaction data has been updated
+          isAwaitingUpdate = false;
+          update();
+
+          Get.to<dynamic>(PaymentAuthorization(this));
+        }
+        break;
+
+      case TransactionStatus.success:
+        // The transaction has been marked as successful on the server-side.
+        if (oldValue.status == TransactionStatus.authorizationRequired) {
+          // The transaction data has been updated
+          isAwaitingUpdate = false;
+          update();
+
```

Comment:
You mentioned this is supposed to a state machine - what did you want to change about this again?

Reply:
Not sure tho... I just think there should be a better way to do this rather than using if else like this. For example, many services in Mojaloop use [`javascript-state-machine`](https://github.com/jakesgordon/javascript-state-machine) to handle the transition.

Reply:
Looks like there's something:
https://pub.dev/packages/state_machine
Maybe we can use this if necessary.

Reply:
Yeah, not really sure though about that since I'm not familiar with the library... but I think it will also introduce some extra complexity to the code, I think shouldn't be a big issue to keep things simple for now